[MJ]미션4: 모든 기물 배치하기

MJbae opened the pull request:
안녕하세요~ MJ입니다.

받은 리뷰에 대해 고민하고 코드에 반영하기를 반복하면서 많이 배우고 있습니다! 이번 리뷰도 잘 부탁드립니다!

### **고민사항**

1. 변경 가능성(또는 확장 가능성)에 대한 여지를 남기는 방법

    **'변경 가능성에 대한 여지' 이해하기**

    - 미션3에 리뷰 중 가장 인상적이었던 것은 확장 가능성이나 변경 가능성에 대한 여지를 코드에 남겨야 한다는 것이었습니다.
    - '변경 가능성에 대한 여지'에 대한 포인트는 '서비스의 변경사항을 코드에 반영할 때, 변경해야 할 지점을 쉽게 파악할 수 있어야 한다는 것'으로 이해하고 있습니다.
    - 이러한 포인트에 대한 프로그래밍적 수단은 예외를 발생시키는 것입니다.
    - 예를 들어, 어떤 이유로 국제 체스 규칙이 변경되어 기물의 색 중에서 흰색이 오렌지색으로 교체되었습니다. 이러한 변경사항을 코드 일부에 반영할 때, 프로그램 곳곳에서 예외가 발생한다면 변경된 체스 규칙을 프로그램 전체에 쉽게 반영할 수 있을 것입니다.

    **'변경 가능성에 대한 여지'를 미션에 적용하기**

    - 아래의 코드는 기물의 색깔과 타입에 대한 매개변수를 받아서 맵핑되는 표현문자를 반환하는 메소드입니다.
    - 변경 전의 코드는 삼항연산자를 활용해서 비교적 간결하게 작성했습니다. 하지만 만약 체스 규정 변경에 따라 흰색 대신 오렌지색을 사용한다고 했을 때, 아래의 코드는 모두 검은색 기물의 표현문자를 반환할 것입니다.
    - 변경 후의 코드는 흰색과 검은색에 대해 모두 조건문으로 확인합니다. 그리고 그 외의 색에 대해서는 예외를 발생시켜서 아래 코드가 변경지점이라는 것을 쉽게 알 수 있습니다.

    ```java
    public String getRepresentation(Color color, Type type) {
        String result = representationMap.get(Color.WHITE.name() + type.name());

        return (color == Color.WHITE) ? result : result.toUpperCase();
    }

    // to

    public String getRepresentation(Color color, Type type) {
        String result = representationMap.get(Color.WHITE.name() + type.name());

        if (color == Color.WHITE) {
            return result;
        } else if (color == Color.BLACK) {
            return result.toUpperCase();
        } else {
            throw new IllegalArgumentException("color is White/Black else");
        }
    }

    ```

### **변경사항**

1. **Feat: StringUtil::appendNewLine, 개행문자 추가 메소드 구현**
    - 변경가능(mutable)한 StringBuilder 객체를 활용하여, 매개변수로 전달받은 문자열(input)에 문자열(NEWLINE)을 추가해도 새로운 문자열 객체를 생성하지 않습니다.
2. **Refactor: Pawn → Piece**
    - Pawn 객체를 Piece 객체로 리팩토링할 때, VO 특성 세 가지 중 immutability(불변성)과 self validation(자기 유효성 검사) 반영했습니다.
    - Value Equality(값 동등성) 특성은 본 미션에 사용되지 않으므로 반영하지 않았습니다.
    - immutability: 생성자에 접근제어자를 private으로 설정하고, 팩토리 메소드로 Piece 객체 생성합니다.
    - self validation: 생성자에 세 가지 속성(color, type, representation)에 대해 유효성 검사 메소드 추가했습니다.
3. **Feat: 표현문자 객체, RepresentationMap 구현**
    - 표현문자를 객체로 구현했습니다.
    - Color와 Type의 이름의 조합을 Key, 표현문자를 Value로 정의했습니다.
    - init() 호출에 따라 표현문자를 맵에 삽입하고 getRepresentation()로Key를 전달 받아서 맵핑되는 표현문자를 반환합니다.
4. **Feat: 모든 기물 생성**
    - switch() 활용해서 기물의 종류에 따라 팩토리 메소드로 해당 기물을 필요한만큼 반복하여 생성합니다
5. **Test: 기물의 초기 위치상태가 정상적으로 표현되는지 테스트**
    - 기물의 초기 위치상태에 대한 문자열 변수를 선언하고, Board::getLocationOfPieces 호출 결과와 앞의 변수의 값이 일치하는 지 확인합니다.
6. **Test: isWhite(), isBlack() 메소드에 대한 테스트**
    - switch문을 활용하여 모든 팩토리 메소드에 대해 isWhite() or isBlack()의 값이 제대로 반환되는지 확인합니다.

Commit:
Refactor: 출력 기능을 Board class에서 Console class로 이동

출력도 기능이므로 하나의 객체에서 전담하는 것이 코드 관리면에서 유리함
추후 유지보수면에서 이러한 장점은 더욱 빛을 발할 것

Commit:
Add: 체스말의 종류에 대한 타입을 설정하는 열거형 추가

Commit:
Refactor: Pawn 객체의 이름을 Piece로 변경

Commit:
Test: Piece class에 팩토리 메소드 추가에 따른 테스트 추가

팩토리 메소드에 반영에 따라 Piece 객체의 인스턴스는 new 키워드가 아니라 팩토리 메소드를 활용함
팩토리 메소드에 따라 인스턴스를 생성할 때 입력하는 Color, Type, representation값이 의도대로 적용됐는지 확인함

Commit:
Refactor: 맥토리 메소드 도입

폰 타입에 대한 팩토리 메소드 추가
팩토리 메소드에 필요한 생성자만 남기고 다른 생성자 불필요에 따라 제거

Commit:
Refactor Test: 체스말의 색, 타입, 표현문자를 테스트하는 부분을 메소드로 분리

테스트의 가독성을 높이기 위해 주요 테스트 부분을 메소드로 분할함
assertAll 활용하여 각각의 테스트의 성공과 실패를 확인할 수 있음

Commit:
Add: 문자열 구분 유틸 추가

OS에 따라 개행문자가 다른 경우가 있음.
OS에 관계없이 동일한 개행문자를 삽입하기 위해 개행문자 추가하는 기능의 메소드 구현

Commit:
Refactor: 지역변수에 final 설정

Commit:
Refactor: appendNewLine()의 반환값을 StringBuilder로 변경

본 메소드는 주로 StringBuilder[] cellsOfBoard에 문자열 삽입할 때 사용될 것
주로 사용되는 데이터구조의 타입에 맞춰서 반환값 변경

Commit:
Test: 체스판 초기화 시 폰의 생성 개수를 체크하는 테스트 생성

Commit:
Test: 폰 생성 개수 확인 테스트의 inline int를 지역변수로 수정

Commit:
Test: 체스판 생성에 따라 모든 기물의 개수 및 초기 출력상태 확인 테스트 추가

Commit:
Refactor Test: createPiece 테스트의 디스플레이 이름 변경

Commit:
Add: showBoard 메소드 추가

체스판의 모든 기물의 위치를 문자열로 반환함
반환된 문자열은 Console::print에서 출력

Commit:
Chore: TODO: 해쉬활용해서 표현문자를 값으로 받아오는 아이디어

Commit:
Refactor: Board:showBoard의 반환값을 그대로 출력

기존에는 Board의 문자열 배열을 가져와서 순회문을 통해 요소요소를 출력함
Board에서 하나의 문자열로 받아와서 그대로 출력하는 방식으로 변경

Commit:
Refactor: 불필요 코드 삭제

Commit:
Add: 표현문자 처리 클래스 추가

기존에는 Piece class 안에 상수로 표현문자를 처리함
Color와 Piece만으로 표현문자를 처리할 수 있을 것이라 판단
color + type을 키값으로 갖고 표현문자를 매칭 값으로 갖는 맵 자료구조 구현
클래스 생성자에 init() 매소드 추가하여 모든 표현문자에 대한 키밸류 값을 초기화함

Commit:
Refactor: 상수로 관리하던 표현문자에 Map자료구조로 관리

RepresentationMap 클래스를 생성하여 내부에 정의한 Map 자료 구조로 표현문자를 관리함

Commit:
Chore: 불필요한 import 제거

Commit:
Refactor: 생성자에 접근제어자 추가

Commit:
Chore: 불필요 변수 및 메소드 제거

Commit:
Refactor Test: 표현문자에 대한 Map 자료구조를 테스트 메소드에 적용

Commit:
Refactor Test: RepresentationMap 도입에 따라 테스트 코드에 적용

기존에는 createPice()에 매개변수로 Piece의 상수로 존재하던 표현문자를 전달했음
RepresentationMap 도입에 따라 기존 상수로 전달하던 부분을 제거함
RepresentationMap의 메소드를 활용하여 실제값과 비교하는 값을 전달

Commit:
Chore: 불필요 변수 제거

Commit:
Add: 모든 기물에 대한 표현문자 삽입

Representation::init에서 모든 기물에 대해 색깔별로 표현문자 값을 맵핑

Commit:
Add: 모든 기물에 대한 생성 팩토리 메소드 추가

Commit:
Refactor: 표현문자의 타입을 문자열로 변경

체스판에 표현문자 표기 시, 문자로 작업하는 것보다 문자열로 하는 것이 가독성면에 이점이 있다고 판단

Commit:
Add: 폰을 포함한 모든 기물의 초기위치를 문자열로 표현하는 기능 추가

getRowOfPieces()에 초기 기물의 위치에 따라 표현문자를 삽입한 것을 반환함
initCellsOfBoard()에 cellsOfBoard의 첫번째와 마지막 열에 폰을 제외한 기물의 표현문자의 값을 할당함
흰색 기물과 검은색 기물 사이의 공백은 마침표로 표현함

Commit:
Test Refactor: 모든 기물의 초기위치를 문자열로 표현하는 기능에 대한 테스트 추가

모든 기물의 초기 위치를 표현하는 기능 추가에 따라 폰의 초기 위치만 테스트하는 코드는 제거

Commit:
Refactor: 접근제어자 수정

Commit:
Refactor: Board의 불필요한 인스턴스 반복 생성 가능성 제거

startGame() 내부에 board를 생성하면 start입력마다 이를 생성함
처음에 한번 생성하는 것이 반복 생성 가능성 제거

Commit:
Refactor: try-with-resources 적용

scanner.close()의 경우 갑작스러운 프로그램 종료 시 작동하지 않음

Commit:
Refactor: self-validity method를 생성자에 추가

VO의 특성 중 하나의 self validity를 구현하기 위해 생성자에 유효성 검사 메소드 추가

Commit:
Refactor: 표현문자에 대한 유효성 검사 추가

RepresentationMap class 내 containsValue 메소드 추가
Piece class 내 생성자에 위의 메소드를 활용한 유효성 검사 메소드 추가

Commit:
Test: 모든 기물에 대해 흑백 상태 확인 테스트 추가

Piece::isBlack, isWhite 추가 전에 해당 메소드에 대한 테스트 케이스 작성
switch문과 Type.values() 활용하여 모든 기물에 대해 isWhite or isBlack 메소드에 대한 테스트 코드 작성

Commit:
Add: isWhite, isBlack 메소드 추가

Commit:
Refactor: isBalck, isWhite 접근제어자 변경

오직 클래스 내부 및 테스트 케이스에만 사용되므로 default로 설정

Commit:
Chore: 불필요한 세미콜론 제거

Commit:
Refactor: 가독성 좋은 메소드명으로 변경

printCellsOfBoard() -> printLocationOfPieces()
showBoard() -> getLocationOfPieces()
위의 메소드명 변경사항을 테스트 코드에 반영

Commit:
Refactor: stream의 중개연산자 변경

기존에서 필터와 맵 중개연산을 두번씩 사용함.
한 번 사용으로 충분히 줄일 수 있음
각각 한번씩 사용하는 것으로 변경:

Commit:
Chore: 불필요 세미콜론 제거 및 메소드 간 공백문자 추가

Commit:
Test: 실제값과 기대값의 비교가 boolean 방식에 따라 isTrue() 사용

Commit:
Refactor: 클래스명 변경(StringUtils -> StringUtil)

Commit:
Refactor: List pieces -> List whitePieces, List blackPieces

기존의 List piecesf를 색깔에 따라 두 개의 리스트로 나누면 enum Color를 매개변수로 전달하지 않아도 기물에 대한 처리가 쉬워짐

Commit:
Refactor: 확장가능성에 대한 여지를 남기기 위해 예외처리

만약 Color의 요소로 ORANGE가 추가된다면, 삼항연산자로 오직 흰색과 검은색에 대해 처리하는 것은 위험함
흰색과 검은색 외의 값에 대처하기 위해 예외처리 추가

Commit:
Refactor: StringBuilder 자료형은 메소드 내부 로직에만 사용, 반환값으로 사용 X

메소드 간 데이터를 주고 받을 때, 보편적으로 많이 사용하는 String을 사용
다른 프로그램과의 호환을 고려하면 String 사용이 적절

Commit:
Test: 예상값이 너무 긴 경우, 지역변수로 사용

board.getLocationOfPieces()에 대한 예상 값이 너무 길어서 initialLocationOfPieces 지역변수를 가져오는 방식으로 변경

Commit:
Refactor: getLocationOfPieces()의 내부로직을 간단하게 변경

기존에는 StringBuilder와 반복문 활용했지만 String::join() 활용

Commit:
Refactor: RepresentationMap::init 중복코드 제거

표현문자 중 대문자의 경우 getRepresnetation에서 color에 조건을 달아서 반환가능
위의 조건을 추가하고 init()에서 대문자 키-밸류 쌍을 제거함

Commit:
Refactor: 불필요한 변수 선언 제거

toLowerCase() 적용하고 변경된 변수를 할당한 변수를 제거하고 메소드 자체를 활용함

Commit:
Chore: 코드 포맷팅 및 불필요 변수 제거

wheejuni commented on src/main/java/mj/chess/Board.java:
```diff
@@ -1,82 +1,133 @@
-        for (Color color : Color.values()) {
-            initPawns(color);
-            initCellsOfBoard(color);
+        switch (type) {
+            case PAWN:
+                piece = (color == Color.WHITE) ? Piece.createWhitePawn() : Piece.createBlackPawn();
+                for (int i = 0; i < 8; i++) this.addPiece(piece);
+                break;
+            case ROOK:
+                piece = (color == Color.WHITE) ? Piece.createWhiteRook() : Piece.createBlackRook();
+                for (int i = 0; i < 2; i++) this.addPiece(piece);
+                break;
+            case KNIGHT:
+                piece = (color == Color.WHITE) ? Piece.createWhiteKnight() : Piece.createBlackKnight();
+                for (int i = 0; i < 2; i++) this.addPiece(piece);
+                break;
+            case BISHOP:
+                piece = (color == Color.WHITE) ? Piece.createWhiteBishop() : Piece.createBlackBishop();
+                for (int i = 0; i < 2; i++) this.addPiece(piece);
+                break;
+            case QUEEN:
+                piece = (color == Color.WHITE) ? Piece.createWhiteQueen() : Piece.createBlackQueen();
+                for (int i = 0; i < 1; i++) this.addPiece(piece);
+                break;
+            case KING:
+                piece = (color == Color.WHITE) ? Piece.createWhiteKing() : Piece.createBlackKing();
+                for (int i = 0; i < 1; i++) this.addPiece(piece);
+                break;
+            default:
+                throw new IllegalStateException("Unexpected value: " + type);
```
이 `switch` 문은 제거해주시고, 앞으로도 가급적 `switch-case` 문은 사용하지 않는 것을 원칙으로 해주세요. 
혹은 제가 파악하지 못한 `switch-case` 문을 사용하는 것의 설계상 이점이 있다면 꼭 알려주세요.

MJbae replied:
@wheejuni 불필요한 switch문을 전부 걷어냈습니다. 찾아보니까 switch문을 일종의 레거시로 보는 분들이 많더군요! 설계 단계에서 신경써서 switch문을 남발하는 습관을 고쳐야겠습니다.

wheejuni commented on src/main/java/mj/chess/Board.java:
```diff
@@ -1,82 +1,133 @@
-        } else {
-            cellsOfBoard[SEVENTH_ROW] = getRowOfPawns(color);
-        }
+    int countPieces() {
```
메소드에 접근제어자가 생략되었네요.

MJbae replied:
접근제어자를 생략한 이유는 현재 상태에서 countPieces()를 사용하는 곳이 테스트코드(BoardTest) 외에는 없기 때문에 package-private 접근제어자를 생략하여 설정한 것입니다.

혹시 이런 상황(특정 메소드를 테스트코드 외에는 사용하지 않는)에 package-private을 생략하는 방법 말고, 더 나은 방법이 있을까요?

kyupid replied:
protected나 default로 구체적으로 안정하고 private 아니면 public으로 설정해주는 게 좋은가용?

MJbae replied:
@kyu-kim-kr 접근제어자는 프로그램의 동작에 이상이 없는 한, 최대한 private에 가깝게 설정하는 것이 정보은닉면에서 좋다고 이해하고 있어요~

wheejuni replied:
@MJbae 테스트코드 이외에 쓰이는 곳이 없는 메소드가....존재해야 할 이유는 무엇일까요? 
근본적인 질문을 다시 한 번 던져보셨으면 좋겠어요. 
아마 요구사항을 충족하기 위해서일거라고 생각이 듭니다. 
그렇다면 요구사항에서 기대하는 접근 제어의 수준을 먼저 생각해보시기를 권합니다.

MJbae replied:
@wheejuni 감사합니다. 요구사항에서 특정 메소드를 테스트하라는 것은 다른 객체와의 협력에 사용된다는 의미가 포함되어 있다고 이해했습니다. 따라서 그러한 메소드에는 접근제어자를 public으로 설정해야 합니다.

wheejuni commented on src/main/java/mj/chess/Board.java:
```diff
@@ -1,82 +1,133 @@
+    private static final int SIZE_OF_ROW = 8;
+    private final List<Piece> whitePieces = new ArrayList<>();
+    private final List<Piece> blackPieces = new ArrayList<>();
+    private final String[] cellsOfBoard = new String[SIZE_OF_ROW];
```
왜 이 배열이 선언되었을까 고민을 많이 했었는데, `initCellsOfBoard()` 메소드를 보니 궁금증이 풀리네요.
궁금증이 풀렸지만 좋은 설계라는 말은 아니고, 요구사항을 너무 요구사항 그대로 해석하신 설계로 판단이 됩니다. 
`Board` 의 상태를 문자열의 형태로 표현할 때는 `Board`가 갖고 있는 `List<Piece>` 상태값만 갖고 표현이 되었으면 좋겠고요, 지금 `cellsOfBoard` 와 같이 단순히 문자열 표현만을 위한 또다른 문자열 상태값을 관리하는 것에는 반대합니다. 

`Board` 가 `Piece` 를 관리하는 주체이므로 관리되고 있는 `Piece` 만을 통해 체스판의 상태가 표현되어야 합니다. 
지금은 우리가 단순히 체스판 위에 기물을 올려놓고, 그 기물들을 그대로 문자열로 표현하는 요구사항만을 구현하고 있는데요, 
앞으로 실제 체스 게임이 체스판 위에서 펼쳐질텐데, 그러면 게임 상황에 따라 `cellsOfBoard` 도 계속 변경돼야 할 겁니다. 
그렇다면 `Piece` 가 움직임을 취할 때마다 `List<Piece>` 를 변경하고, 또 `cellsOfBoard` 도 변경해야 하고.... 메소드 호출 한 번에 수차례의 `private` 메소드 호출이 발생하고, 심지어 `private` 메소드간에 강한 결합 관계가 형성되어 리팩토링 하기도, 유닛테스트 하기도 어려운 코드로 변모할 겁니다. 이런 설계를 원하신 건 아닐 겁니다. 

`cellsOfBoard` 를 없애야하는 이유를 장황하게 설명드린 것 같은데요, 큰 설계의 변화가 필요할 것 같아서 조금 조심스럽지만 지금 단계에서 이 변수를 걷어내는 작업이 꼭 필요할 것 같아 의견 드립니다.

MJbae replied:
@wheejuni cellsOfBoard를 제거하고 List<Piece>만으로 다시 구현했습니다.  Board 클래스 자체가 Piece 클래스에 대해 관리하는 책임을 갖기 때문에 cellsOfBoard와 같이 전체 Piece에 대해 관리하는 또다른 주체는 불필요합니다. 요구사항만 보고 무턱대고 구현부터 했기 때문에 이러한 문제점이 생겼습니다. 전체적인 로직에 더 신경쓰고 구현하는 습관을 가져야겠습니다.

wheejuni commented on src/main/java/mj/chess/Board.java:
```diff
@@ -1,82 +1,133 @@
-    public void add(Pawn pawn) {
-        pawns.add(pawn);
+    private void addPiece(Piece piece) {
+        if (piece.isWhite()) {
+            whitePieces.add(piece);
+        }
+        if (piece.isBlack()) {
+            blackPieces.add(piece);
+        }
```
~~`Piece` 는 흑이거나 백일 거라고 생각됩니다. 또 앞으로의 요구사항에도 거의 100%의 확률로 색깔이 추가될 여지는 없어 보여요. 
그렇다면 아래와 같이 간소화할 수 있을 것 같네요.~~

PR 메세지를 뒤늦게 읽어서 왜 이런 방향의 구현을 하셨는지 이해했습니다. 
이에 대한 저의 생각은 조금 정리해서 다시 알려드릴게요.

```suggestion
        if (piece.isWhite()) {
            whitePieces.add(piece);
        }
        blackPieces.add(piece);

MJbae replied:
'변경사항에 대한 여지'를 모든 코드에 일일이 반영하는 것은 전체 코드의 간결성을 크게 훼손하는군요!  

wheejuni의 포인트는 서비스의 변경사항이 발생해도, 코어한 일부 코드의 변경만으로 프로그램 전체에 서비스 변경사항이 반영되어야 한다는 것으로 이해했습니다.
특히 '설계적 이점'이라는 키워드를 계속 강조하신 이유가 설계 단계에서 이러한 부분이 반영되어야 한다는 것이구요.

wheejuni commented on src/main/java/mj/chess/Board.java:
```diff
@@ -1,82 +1,133 @@
+        }
+
+        return pieces.stream()
+                .filter(piece -> piece.getType() == Type.PAWN && piece.getColor() == color)
```
`piece.getColor() == color` 와 같이 `Piece` 에게 색깔을 물어봐 체크하는 것은 정말 좋은 설계로 생각됩니다. 
그렇다면 앞의 `else if` 문을 제거할 수 있는 힌트를 얻을 수 있을 것 같은데요. 
`blackPieces` 와 `whitePieces` 를 합친 뒤 기물들에게 일일이 색깔을 물어봐서 일치 여부를 판정하는 로직은 어려울까요?
그렇다면 `else if` 도 걷어낼 수 있을 것 같은데요. 

앞으로 과제 진행에서 `else` 키워드도 사용을 최소화하는 것을 제 1의 원칙으로 삼아주세요.

wheejuni commented on src/main/java/mj/chess/Console.java:
```diff
@@ -3,37 +3,43 @@
 import java.util.Scanner;
 
 public class Console {
+    private Board board;
```
`main` 문을 벗어나는 게임이 존재할 수 있을까요?
실행시 최초 한 번 인스턴스가 생성되고, 그 이후에 다른 곳에서 엑세스될 필요가 없다면 `main` 문의 로컬 변수로 선언하고 관리하는 것으로 충분하겠다는 생각은 듭니다.

wheejuni commented on src/main/java/mj/chess/Console.java:
```diff
@@ -3,37 +3,43 @@
 
-        scanner.close();
+        String command;
+        board = new Board();
```
여기서 아래와 같이 선언해주는게 나아 보인다는 말씀이었습니다. 

```suggestion
        Board board = new Board();
```

만약 아래 `private` 메소드들때문에 그렇게 설계하지 못하셨다면, `private` 메소드들에서 `Board` 타입의 파라메터를 받아주면 될 것으로 생각합니다.

MJbae replied:
그렇군요! 불필요하게 클래스 멤버변수를 만들어서 자원을 낭비하고 있다는 말씀으로 이해했습니다.

클래스 멤버변수는 스태틱영역(메소드영역)에 계속 남지만, 메소드의 지역변수는 해당 메소드가 스택영역에서 pop되면 함께 제거되기 때문에 자원 관리면에서 더 낫겠네요.

wheejuni replied:
@MJbae 자원 관리보다 더 중요한 문제들이 있습니다. 다만 지금 단계에서 다루기는 조금 깊은 내용들이고요, 
단계를 진행해가면서 하나씩 발견해 보시는것도 큰 재미가 될 것 같아서 그 재미를 지금 망치지는 않을게요. 
_왜 그때 그 리뷰어는 이런 말을 했을까?_ 를 항상 유념해 보세요.

MJbae replied:
코드스쿼드의 이런 문화(답을 알려주지 않고 직접 찾게 만드는)는 지적 호기심을 많이 자극하네요ㅎㅎ
본 리뷰는 항상 유념하고 있겠습니다!

wheejuni commented on src/main/java/mj/chess/pieces/Piece.java:
```diff
@@ -0,0 +1,121 @@
+
+    public static Piece createWhiteRook() {
+        return new Piece(Color.WHITE, Type.ROOK);
+    }
+
+    public static Piece createBlackRook() {
+        return new Piece(Color.BLACK, Type.ROOK);
+    }
+
+    public static Piece createWhiteBishop() {
+        return new Piece(Color.WHITE, Type.BISHOP);
+    }
+
+    public static Piece createBlackBishop() {
+        return new Piece(Color.BLACK, Type.BISHOP);
+    }
+
+    public static Piece createWhiteQueen() {
+        return new Piece(Color.WHITE, Type.QUEEN);
+    }
+
+    public static Piece createBlackQueen() {
+        return new Piece(Color.BLACK, Type.QUEEN);
+    }
+
+    public static Piece createWhiteKnight() {
+        return new Piece(Color.WHITE, Type.KNIGHT);
+    }
+
+    public static Piece createBlackKnight() {
+        return new Piece(Color.BLACK, Type.KNIGHT);
+    }
+
+    public static Piece createWhiteKing() {
+        return new Piece(Color.WHITE, Type.KING);
+    }
+
+    public static Piece createBlackKing() {
+        return new Piece(Color.BLACK, Type.KING);
+    }
```
필수 사항은 아닙니다만, 이 스태틱 팩토리 메소드들은 별도의 클래스에서 관리하는 것도 나쁘지 않아보입니다.

MJbae replied:
코드의 간결성면에서 제안하신 방법을 반영하는 것이 좋겠네요~

wheejuni commented on src/main/java/mj/chess/pieces/Piece.java:
```diff
@@ -0,0 +1,121 @@
+            }
+        }
+
+        if (!isValid) throw new IllegalArgumentException("Invalid Type");
```
`if` 문에는 항상 중괄호 `{}` 를 사용하는 것을 원칙으로 해주시면 좋겠습니다.

MJbae replied:
이 부분은 자바의 코딩 컨벤션이군요. 컨벤션을 다시 읽어봐야겠네요.

wheejuni commented on src/main/java/mj/chess/pieces/RepresentationMap.java:
```diff
@@ -0,0 +1,37 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class RepresentationMap {
```
이 클래스의 존재 필요성에 대해... 조금 다시 생각해 보셨으면 하는데요, 
우리는 `Color` 와 `Type` 만 있으면 기물의 표현기호도 자연스레 얻을 수가 있습니다. 
방법은 여러가지가 있겠는데요, `Type` 에 흑, 백 각 경우에 대한 표현기호를 보관해도 되고요,
혹은 `Type` 에 베이스 표현기호를 둔 후 `Color` 타입 파라메터를 받는 메소드를 통해 각 색깔별 표현기호를 되돌려주는 로직을 설계할 수도 있습니다. 그러니까 아래와 같은 구현이 가능합니다. 

```java
public enum Type {
....

    private String getRepresentation(Color color) {
    .....
```

그래서 이 `RepresentationMap` 은 다소 불필요하다고 생각되는데요,
이 클래스를 제거할 방법을 찾아보시고 설계를 다시 해보시기를 강력하게 권합니다.

MJbae replied:
그렇군요! 고민해보겠습니다.

MJbae replied:
@wheejuni 제안하신 방법으로 수정했습니다. 기존 객체를 이용하는 것이 코드의 재사용에 따라 성능을 향상시킬 수 있습니다. 더불어 요구사항 중 '특정 타입과 색깔에 해당하는 표현문자를 출력하라'에 대해 Type 자체적으로 책임을 충분히 질 수 있습니다.

wheejuni reviewed (changes requested):
4단계까지 진행하시느라 고생 많으셨습니다. 👍 💯 
고생 많이 하셨는데 설계에서 변경 포인트를 많이 지적드려서 조금 죄송하기도 하네요. 
많은 변경이 필요할 것으로 생각됩니다. 개선의 방향은 코멘트를 통해 남겨 드렸는데요,
동의되지 않는다거나, 이해가 어려우신 부분은 언제든지 알려주시기 바랍니다.

kyupid commented on src/main/java/mj/chess/Board.java:
```diff
@@ -1,82 +1,133 @@
 
-    public int size() {
-        return pawns.size();
+    void initialize() {
+        for (Type type : Type.values()) {
+            for (Color color : Color.values()) {
+                initPieces(color, type);
```
타입별로 컬러별로 아예 빼버리고 이런식으로 초기화하니까 훨씬 깔끔하네요 ㅎㅎ

Commit:
Refactor: Board 객체의 인스턴스 생성 부분을 Console::runConsole 내부로 이동

기존에는 Console의 클래스멤버변수로 사용되었음.
자원 관리면에서 클래스변수는 스태틱 영역에 남아있지만 runConsole()의 지역변수로 사용되면
해당 메소드가 호출스택에서 pop되면 함께 제거됨.
결과적으로 자원관리면에서 위와 같은 리팩토링을 진행

Commit:
Refactor: 표현문자를 얻는 로직을 RepresentationMap -> Type 이동

다음의 설계적 요구, '특정 타입과 색깔에 해당하는 표현문자를 출력하라'에 대해 새로운 객체(RepresentationMap)가 책임을 졌음
하지만 기존에 존재하는 객체, Type이 위의 설계적 요구에 대한 책임을 충분히 질 수 있었음
기존 객체를 이용하는 것이 코드의 재사용에 따라 성능을 향상시킬 수 있으므로 Type 객체에서 표현문자에 대한 책임을 지는 것으로 수정

Commit:
Refactor: 표현문자 관련 로직 이동(->Type)과 연관된 부분 수정

테스트 코드 내 representationMap 인스턴스 활용하는 부분 제거
RepresentationMap 객체 제거
Piece와 Board 내 Representation 활용하는 부분 제거하고 Type::getRepresentation으로 대체

Commit:
Refactor: Piece의 내부클래스 PieceMaker에 팩토리 메소드 이동

Piece 클래스의 VO 특성을 유지하기 위해 PieceMaker 클래스를 내부 클래스로 이동함
프로그램 전반에 걸쳐 이러한 변경사항을 반영함

Commit:
Refactor: 조건문의 본문에 한줄의 코드만 있어도 중괄호 처리

이것은 자바 코딩 컨벤션이므로 따를 것

Commit:
Refactor: Board::initPieces의 로직(switch문 활용 부분 개선)을 Enum Type으로 이동

Board::initPieces의 내부로직에 switch문을 활용함 type에 대한 case에 따라 기물을 생성하고 체스판에 추가시켰음
switch문에 해당하는 부분을 type으로 이동시켰음
함수형 인터페이스 BiConsumer 활용해서 initPiece() 호출에 따라 해당 로직을 실행함

Commit:
Refactor: enum Type 생성 시 해당 타입의 표현문자를 보드에 생성

Commit:
Refactor: Board 클래스 내 불필요한 상태관리 자료구조(String[] cellsOfBoard) 제거

Board 자체가 Piece 객체를 관리하기 위한 주체가 되어야 함
기존 코드에는 Board 객체 내에 Piece의 표현문자(presentation)를 관리하는 또 다른 관리주체가 존재했음
List<Piece> 자료형으로 Board의 부분 구조를 완성하고 이를 통해 Board 객체 내 모든 Piece를 관리함

Commit:
Chore: public 메소드를 상단으로 이동

public 메소드는 다른 객체와의 협력을 표현하기 때문에 객체 간의 협력내용을 파악하기 위해 자주 볼 것이라 판단함
자주 봐야 하는 코드를 상단에 배치하는 것이 코드의 전체적인 흐름을 파악하는데 유리할 것이라 판단

Commit:
Test: 테스트 코드에 불필요한 switch문 제거

Commit:
Refactor: 체스판의 행의 개수에 대해 클래스 멤버변수로 선언

기존에는 하드코딩으로 8이라는 숫자를 직접 주입했음
ROWS라는 변수로 선언하여 사용

Commit:
Refactor: List[] rows 추가해서 각 행에 해당하는 ArrayList<Piece> 관리

List[] rows 도입 목적은 각 행의 ArrayList를 순회하여 중복코드 제거 주로 각 행을 담당하는 ArrayList를 조회하는데 사용함

wheejuni commented on src/main/java/mj/chess/Board.java:
```diff
@@ -1,82 +1,117 @@
+    private final int SEVENTH_ROW = 6;
+    private final int EIGHTH_ROW = 7;
+
+    private final List[] rows = new ArrayList[ROWS];
```
의도는 이해하겠으나 가독성이 너무 좋지 않습니다. 
`List`를 리스트로 담을 수 있습니다. 

```suggestion
    private final List<List<Piece>> rows = new ArrayList();
```

wheejuni reviewed (changes requested):
한 가지 추가 수정사항만 짚고 바로 다음 단계로 넘어갑시다! 👏

MJbae committed:
Refactor: List[] -> List<List<Piece>>

가독성 문제에 따라 List 배열을 List List로 수정

Co-authored-by: Hwijun Jeong <me@wheejuni.com>

MJbae commented:
> 한 가지 추가 수정사항만 짚고 바로 다음 단계로 넘어갑시다! 👏

@wheejuni 제안하신 부분을 따라서 수정했습니다 :)
